# Tarea 1

Problema1:
El objetivo de las bases es que realicemos un modelo de clasificación para poder contestar la siguiente pregunta:

      ¿Damos un crédito bancario o no?

Para empezar con la exploración de la base, tendrán que generar un análisis descriptivo de la información que les compartiré:
 1.- Los campos _id_ que son los identificadores de la información.
 2.- base_ini_interna
 3.- base_ini_parte1
 4.- base_ini_externa
 5.- bases de comportamiento
 6.- base_inversión, es una variable simulada de una gamma, es
 7.- Cuando una variable es categórica, está implícito en el nombre de la base. No tenemos un diccionario, simplemente que son categóricas.
 8.- El resto entonces son continuas. Es decir, los nombre estám enmascarados, ni siquiera yo sé qué significan.
 

¿Cómo hago los cruces? Exploren la información.
¿Qué análisis puedo realizar? Exploren ustedes, apóyense de los comportamientos de los ids.
Pero... ¿Las preguntas son abiertas? Así es, quiero que exploren. Hay bases que vienen en csv y otras en txt. Vean como tratarlas.

El resto de las tarea, son las preguntas que les dejé el dìa miércoles. Les coparto los programas que utilizamos.
Pero resumo un poco:


Problema 2:
De la base data_schools sacar obtener el error estándar con Bootstrap de las variables de la base. 
Recordemos que tienen que simular N Muestras Boostrap con tamño n. Luego, el otro ejercicio es que lo hagan con n-1.

Problema 3: Calcular el eror estándar de las betas del modelo lineal, les expliqué la idea de este problema.

Primero importamos las librerías necesarias

In [1]:
from pylab import*
import sqldf
import numpy as np
import pandas as pd
import scipy.stats 
from scipy.special import gamma
import seaborn as sns
from scipy.stats import spearmanr
from scipy.stats import beta
from sklearn import linear_model

# Pregunta 1

In [2]:
path = "/Users/ivanas.o/Downloads/"
#Cargamos las bases de datos
bd_ini_int = pd.read_csv(path + "base_ini_interna.csv")
bd_ini_pt1 = pd.read_csv(path + "base_ini_parte1.csv")
bd_comp3 = pd.read_csv(path + "base_comportamiento_3.csv")
# Archivos .txt
bd_comp2 = pd.read_csv(path + "base_comportamiento_2.txt", sep = " ")
bd_ini_ext = pd.read_csv(path+"base_ini_externa.txt",sep = " ")
bd_inver = pd.read_csv(path + "base_inversion.txt", sep = " ")

Empezamos el análisis exploratorio de las bases de datos. 

### base_ini_interna

In [3]:
bd_ini_int = bd_ini_int.sort_values(by = ["fecha_inicio"])
bd_ini_int.head()

_id_ fecha_inicio  comp_interno1 comp_interno2  comp_interno3  \
17236  2333   2016-01-04    6079.027356          b'1'           30.0   
47365   488   2016-01-04     911.854103          b'0'           19.0   
21961  1964   2016-01-04       0.000000          b'1'           19.0   
49384  1220   2016-01-04    2482.269504          b'0'            9.0   
16983  1226   2016-01-04       0.000000          b'1'            3.0   

       comp_interno4  categoria1  categoria2  categoria3  
17236           50.0         2.0         9.0       175.0  
47365           48.0         3.0         3.0       164.0  
21961           53.0         4.0         3.0       164.0  
49384           70.0         3.0         9.0       154.0  
16983           75.0         3.0         9.0       148.0

In [4]:
bd_ini_int.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50000 entries, 17236 to 32134
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   _id_           50000 non-null  int64  
 1   fecha_inicio   50000 non-null  object 
 2   comp_interno1  50000 non-null  float64
 3   comp_interno2  50000 non-null  object 
 4   comp_interno3  50000 non-null  float64
 5   comp_interno4  50000 non-null  float64
 6   categoria1     50000 non-null  float64
 7   categoria2     50000 non-null  float64
 8   categoria3     50000 non-null  float64
dtypes: float64(6), int64(1), object(2)
memory usage: 3.8+ MB


In [5]:
bd_ini_int.nunique()

_id_             50000
fecha_inicio      1557
comp_interno1      680
comp_interno2        2
comp_interno3       32
comp_interno4       55
categoria1           6
categoria2          13
categoria3          32
dtype: int64

Notamos que no tenemos valores nulos en id . También, obtenemos el número de valores distintos que puede tener cada variable (columna) en la base de datos. 

Como todos los id son distintos y tenemos en total el mismo número de id que el total de id que es de 50000, lo que implica que no hay id repetidos. 

Por otro lado, las demás variables tienen un número menor a 50000 de valores distintos, es decir, hay valores que se repiten. 

### base_ini_externa

In [6]:
bd_ini_ext.head()

_id_  comp_externo1  comp_externo2  comp_externo3  comp_externo4
0  78213     260.869565      25.854639           56.0         -230.0
1  76457      40.579710       2.872738           62.0         -230.0
2  40666            NaN            NaN          119.0         -230.0
3   8329     349.275362      34.616490           94.0         -230.0
4  68389     276.811594      27.434645          125.0         -230.0

In [7]:
bd_ini_ext.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   _id_           50000 non-null  int64  
 1   comp_externo1  47831 non-null  float64
 2   comp_externo2  44410 non-null  float64
 3   comp_externo3  49994 non-null  float64
 4   comp_externo4  50000 non-null  float64
dtypes: float64(4), int64(1)
memory usage: 1.9 MB


In [8]:
bd_ini_ext.nunique()

_id_             50000
comp_externo1     7533
comp_externo2     6982
comp_externo3      237
comp_externo4        6
dtype: int64

Del mismo modo, como en la base anterior, tenemos 50,000 id  distintos que es igual al total de id, es decir, que no se repiten. 

Por otro lado, los componentes externos 1,2 y 3 tienen valores nulos.

### base_ini_parte1

In [9]:
bd_ini_pt1.head()

_id_    _id2_  comp_mixto1
0  78213  1798899      0.33522
1  76457  1758511      0.33334
2  40666   935318      0.29186
3   8329   191567      0.29112
4  68389  1572947      0.27318

In [10]:
bd_ini_pt1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   _id_         50000 non-null  int64  
 1   _id2_        50000 non-null  int64  
 2   comp_mixto1  50000 non-null  float64
dtypes: float64(1), int64(2)
memory usage: 1.1 MB


In [11]:
bd_ini_pt1.nunique()

_id_           50000
_id2_          50000
comp_mixto1     7098
dtype: int64

En esta base tampoco hay entradas nulas y notamos que ambos id no están repetidos. 

### base_comportamiento_2

In [12]:
bd_comp2.head(20)

num_caso      mes_informacion  Tiempo_aparicion
0          1  2016-04-01 00:00:00               0.0
1          1  2016-05-01 00:00:00               1.0
2          1  2016-06-01 00:00:00               2.0
3          1  2016-07-01 00:00:00               3.0
4          1  2016-08-01 00:00:00               4.0
5          1  2016-09-01 00:00:00               5.0
6          1  2016-10-01 00:00:00               6.0
7          1  2016-11-01 00:00:00               7.0
8          1  2016-12-01 00:00:00               8.0
9          1  2017-01-01 00:00:00               9.0
10         1  2017-02-01 00:00:00              10.0
11         1  2017-03-01 00:00:00              11.0
12         1  2017-04-01 00:00:00              12.0
13         1  2017-05-01 00:00:00              13.0
14         1  2017-06-01 00:00:00              14.0
15         1  2017-07-01 00:00:00              15.0
16         1  2017-08-01 00:00:00              16.0
17         1  2017-09-01 00:00:00              17.0
18         1  2017-10-01 00:00:00              18.0
19         1  2017-11-01 00:00:00              19.0

In [13]:
bd_comp2[60:67]

num_caso      mes_informacion  Tiempo_aparicion
60         1  2021-04-01 00:00:00              60.0
61         1  2021-05-01 00:00:00              61.0
62         7  2016-01-01 00:00:00               0.0
63         7  2016-02-01 00:00:00               1.0
64         7  2016-03-01 00:00:00               2.0
65         7  2016-04-01 00:00:00               3.0
66         7  2016-05-01 00:00:00               4.0

In [14]:
bd_comp2.tail(20)

num_caso      mes_informacion  Tiempo_aparicion
1455516     98971  2021-04-01 00:00:00               0.0
1455517     98971  2021-05-01 00:00:00               1.0
1455518     98972  2021-04-01 00:00:00               0.0
1455519     98972  2021-05-01 00:00:00               1.0
1455520     98977  2021-04-01 00:00:00               0.0
1455521     98977  2021-05-01 00:00:00               1.0
1455522     98980  2021-04-01 00:00:00               0.0
1455523     98980  2021-05-01 00:00:00               1.0
1455524     98986  2021-04-01 00:00:00               0.0
1455525     98986  2021-05-01 00:00:00               1.0
1455526     98987  2021-04-01 00:00:00               0.0
1455527     98987  2021-05-01 00:00:00               1.0
1455528     98988  2021-04-01 00:00:00               0.0
1455529     98988  2021-05-01 00:00:00               1.0
1455530     98990  2021-04-01 00:00:00               0.0
1455531     98990  2021-05-01 00:00:00               1.0
1455532     98992  2021-04-01 00:00:00               0.0
1455533     98992  2021-05-01 00:00:00               1.0
1455534     98997  2021-04-01 00:00:00               0.0
1455535     98997  2021-05-01 00:00:00               1.0

Al ver los primeros y úlitmos datos de la base bd_comp2, notamos que por cada número de caso hay varios registros. Para cada número de caso hay un mes de información y tiempo de aparición. Un mismo caso tiene varias apariciones. También que los números de casos no son seriados, es decir, está resigtrado el caso 1 , pero no hay algo al respecto del caso 2 al 6, mientras que para el caso 7 si hay registros. 

Lo mismo sucede en varias partes de la base de datos como se observa en los outputs anteriores. Esto significa que ciertos casos iniciaron un trpamite que otros casos no hicieron. Además, de este trámite se lleva un registro del mes y el tiempo de aparición, lo cual se explicará detalladamente más adelante. 

In [15]:
bd_comp2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1455536 entries, 0 to 1455535
Data columns (total 3 columns):
 #   Column            Non-Null Count    Dtype  
---  ------            --------------    -----  
 0   num_caso          1455536 non-null  int64  
 1   mes_informacion   1455536 non-null  object 
 2   Tiempo_aparicion  1455536 non-null  float64
dtypes: float64(1), int64(1), object(1)
memory usage: 33.3+ MB


No tenemos valores nulos. 

In [16]:
bd_comp2.nunique()

num_caso            49352
mes_informacion        65
Tiempo_aparicion       65
dtype: int64

In [17]:

for col in bd_comp2.columns[1:]:
 print("Valores únicos de "+col+": {0}\n".format(sorted(bd_comp2[col].unique())))
 


Valores únicos de mes_informacion: ['2016-01-01 00:00:00', '2016-02-01 00:00:00', '2016-03-01 00:00:00', '2016-04-01 00:00:00', '2016-05-01 00:00:00', '2016-06-01 00:00:00', '2016-07-01 00:00:00', '2016-08-01 00:00:00', '2016-09-01 00:00:00', '2016-10-01 00:00:00', '2016-11-01 00:00:00', '2016-12-01 00:00:00', '2017-01-01 00:00:00', '2017-02-01 00:00:00', '2017-03-01 00:00:00', '2017-04-01 00:00:00', '2017-05-01 00:00:00', '2017-06-01 00:00:00', '2017-07-01 00:00:00', '2017-08-01 00:00:00', '2017-09-01 00:00:00', '2017-10-01 00:00:00', '2017-11-01 00:00:00', '2017-12-01 00:00:00', '2018-01-01 00:00:00', '2018-02-01 00:00:00', '2018-03-01 00:00:00', '2018-04-01 00:00:00', '2018-05-01 00:00:00', '2018-06-01 00:00:00', '2018-07-01 00:00:00', '2018-08-01 00:00:00', '2018-09-01 00:00:00', '2018-10-01 00:00:00', '2018-11-01 00:00:00', '2018-12-01 00:00:00', '2019-01-01 00:00:00', '2019-02-01 00:00:00', '2019-03-01 00:00:00', '2019-04-01 00:00:00', '2019-05-01 00:00:00', '2019-06-01 00:00:00'

Con lo anterior, corroboramos que el número de valores úncios de num_caso es de 49,352, el cual es menor a el total de registros de num_casos. Esto nos indica que hay num_caso repetidos.

Por otro lado, las variables mes_informacion y Tiempo_aparicion tienen el mismo número de valores únicos, el cual es de 65. 

Además, se observa algo particular. Que para un mismo número de caso, este puede tener 64 apariciones, es de 64 registros. Los cuales están en orden cronológico , ási, mes_informacion es el mes en el que el caso hace aparicion y Tiempo_aparicion es número que se le asigna a dicha aparición dependiendo de su orden cronológico.

Para visualizarlo mejor, voy a acomodar la tabla por Tiempo_aparicion, esto, con la finalidad de identificar cuales numeros de casos llegan a tener 64 apariciones. 

In [18]:
bd_comp2 = bd_comp2.sort_values(by = "Tiempo_aparicion")
bd_comp2

num_caso      mes_informacion  Tiempo_aparicion
0              1  2016-04-01 00:00:00               0.0
989133     45151  2018-05-01 00:00:00               0.0
989170     45156  2018-07-01 00:00:00               0.0
989205     45157  2018-05-01 00:00:00               0.0
989242     45158  2018-04-01 00:00:00               0.0
...          ...                  ...               ...
55435       2095  2021-05-01 00:00:00              64.0
13236        509  2021-05-01 00:00:00              64.0
18801        707  2021-05-01 00:00:00              64.0
12253        488  2021-05-01 00:00:00              64.0
17498        657  2021-05-01 00:00:00              64.0

[1455536 rows x 3 columns]

Ahora, me centraré en el caso 1894 que se encuentra en el renglón 50277 de nuestra base de datos. Lo que se hará a continuación es volver a poner nuestra base bd_comp2 sin acomodar y después veremos la información que está contenida del renglón 50213 al renglón 50277.

¿Qué se espera ver? Por lo comentado anteriormente, se espera que la columna num_caso tenga en todos los renglones el número 1894. Por otro lado, se espera que el mes de información vaya cambiando en orden ascendente. Minetras que el tiempo de aparición contenga la numeración del 0 al 64. 

In [19]:
#Cargamos la base sin acomodar
bd_comp2 = pd.read_csv(path + "base_comportamiento_2.txt", sep = " ")
#Mostramos los renglones 50213 a 50277
bd_comp2[50213:50278]

num_caso      mes_informacion  Tiempo_aparicion
50213      1894  2016-01-01 00:00:00               0.0
50214      1894  2016-02-01 00:00:00               1.0
50215      1894  2016-03-01 00:00:00               2.0
50216      1894  2016-04-01 00:00:00               3.0
50217      1894  2016-05-01 00:00:00               4.0
...         ...                  ...               ...
50273      1894  2021-01-01 00:00:00              60.0
50274      1894  2021-02-01 00:00:00              61.0
50275      1894  2021-03-01 00:00:00              62.0
50276      1894  2021-04-01 00:00:00              63.0
50277      1894  2021-05-01 00:00:00              64.0

[65 rows x 3 columns]

Efectivamente se cumple nuestra hipótesis. A continuación, se va a acomodar a bd_comp2 por mes_informacion. 

In [20]:
#Acomodamos por mes_informacion
bd_comp2 = bd_comp2.sort_values(by = "mes_informacion")
bd_comp2.head()

num_caso      mes_informacion  Tiempo_aparicion
32240      1220  2016-01-01 00:00:00               0.0
44912      1704  2016-01-01 00:00:00               0.0
74499      2824  2016-01-01 00:00:00               0.0
20479       775  2016-01-01 00:00:00               0.0
27608      1049  2016-01-01 00:00:00               0.0

### base_ini_comportamiento3

In [21]:
bd_comp3 = pd.read_csv(path + "base_comportamiento_3.csv")
bd_comp3.head(20)

num_caso      mes_informacion  comportamiento_j
0          1  2016-04-01 00:00:00               0.0
1          1  2016-05-01 00:00:00               0.0
2          1  2016-06-01 00:00:00               0.0
3          1  2016-07-01 00:00:00               0.0
4          1  2016-08-01 00:00:00               0.0
5          1  2016-09-01 00:00:00               0.0
6          1  2016-10-01 00:00:00               0.0
7          1  2016-11-01 00:00:00               0.0
8          1  2016-12-01 00:00:00               0.0
9          1  2017-01-01 00:00:00               1.0
10         1  2017-02-01 00:00:00               0.0
11         1  2017-03-01 00:00:00               0.0
12         1  2017-04-01 00:00:00               1.0
13         1  2017-05-01 00:00:00               1.0
14         1  2017-06-01 00:00:00               1.0
15         1  2017-07-01 00:00:00               0.0
16         1  2017-08-01 00:00:00               0.0
17         1  2017-09-01 00:00:00               0.0
18         1  2017-10-01 00:00:00               0.0
19         1  2017-11-01 00:00:00               0.0

In [22]:
bd_comp3[60:65]

num_caso      mes_informacion  comportamiento_j
60         1  2021-04-01 00:00:00               0.0
61         1  2021-05-01 00:00:00               0.0
62         7  2016-01-01 00:00:00               0.0
63         7  2016-02-01 00:00:00               0.0
64         7  2016-03-01 00:00:00               0.0

In [23]:
bd_comp3.tail(20)

num_caso      mes_informacion  comportamiento_j
1455516     98971  2021-04-01 00:00:00               0.0
1455517     98971  2021-05-01 00:00:00               0.0
1455518     98972  2021-04-01 00:00:00               0.0
1455519     98972  2021-05-01 00:00:00               0.0
1455520     98977  2021-04-01 00:00:00               0.0
1455521     98977  2021-05-01 00:00:00               0.0
1455522     98980  2021-04-01 00:00:00               0.0
1455523     98980  2021-05-01 00:00:00               0.0
1455524     98986  2021-04-01 00:00:00               0.0
1455525     98986  2021-05-01 00:00:00               0.0
1455526     98987  2021-04-01 00:00:00               0.0
1455527     98987  2021-05-01 00:00:00               0.0
1455528     98988  2021-04-01 00:00:00               0.0
1455529     98988  2021-05-01 00:00:00               0.0
1455530     98990  2021-04-01 00:00:00               0.0
1455531     98990  2021-05-01 00:00:00               0.0
1455532     98992  2021-04-01 00:00:00               0.0
1455533     98992  2021-05-01 00:00:00               0.0
1455534     98997  2021-04-01 00:00:00               0.0
1455535     98997  2021-05-01 00:00:00               0.0

Para cada número de caso hay un mes de informacipoin y un comportamiento j en dicho mes de información. Además, se observa algo interesante. Al ver los primeros 65 datos de bd_comp3 y compararlos con los primeros 65 datos de bd_comp2, parece ser que los renglones correpsonden al mismo número de caos con el mismo mes de información. También observamos lo mismo de las colas de dichas bases de datos. 

Por otro lado, el bd_comp3 parece ser que se salta los mismos numeros de casos que en bd_comp2, es decir, por ejemplo, en bd_comp2 no había registros correpsondientes a los casos 2,3,4,5,6 y parece ser que en bd_comp3 tampoco. Más adelante haremos unas consultas para corroborar las observaciones que se plantean en este párrafo acerca si tienen o no los mismos números de casos.

In [24]:
bd_comp3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1455536 entries, 0 to 1455535
Data columns (total 3 columns):
 #   Column            Non-Null Count    Dtype  
---  ------            --------------    -----  
 0   num_caso          1455536 non-null  int64  
 1   mes_informacion   1455536 non-null  object 
 2   comportamiento_j  1455536 non-null  float64
dtypes: float64(1), int64(1), object(1)
memory usage: 33.3+ MB


In [25]:
bd_comp3.nunique()

num_caso            49352
mes_informacion        65
comportamiento_j       18
dtype: int64

In [26]:
bd_comp3["comportamiento_j"].unique()

array([ 0.,  1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9., 13., 10., 12.,
       16., 18., 14., 15., 11.])

In [27]:
for col in bd_comp3.columns[1:]:
    print("Valores únicos de "+col+": {0}\n".format(sorted(bd_comp3[col].unique())))

Valores únicos de mes_informacion: ['2016-01-01 00:00:00', '2016-02-01 00:00:00', '2016-03-01 00:00:00', '2016-04-01 00:00:00', '2016-05-01 00:00:00', '2016-06-01 00:00:00', '2016-07-01 00:00:00', '2016-08-01 00:00:00', '2016-09-01 00:00:00', '2016-10-01 00:00:00', '2016-11-01 00:00:00', '2016-12-01 00:00:00', '2017-01-01 00:00:00', '2017-02-01 00:00:00', '2017-03-01 00:00:00', '2017-04-01 00:00:00', '2017-05-01 00:00:00', '2017-06-01 00:00:00', '2017-07-01 00:00:00', '2017-08-01 00:00:00', '2017-09-01 00:00:00', '2017-10-01 00:00:00', '2017-11-01 00:00:00', '2017-12-01 00:00:00', '2018-01-01 00:00:00', '2018-02-01 00:00:00', '2018-03-01 00:00:00', '2018-04-01 00:00:00', '2018-05-01 00:00:00', '2018-06-01 00:00:00', '2018-07-01 00:00:00', '2018-08-01 00:00:00', '2018-09-01 00:00:00', '2018-10-01 00:00:00', '2018-11-01 00:00:00', '2018-12-01 00:00:00', '2019-01-01 00:00:00', '2019-02-01 00:00:00', '2019-03-01 00:00:00', '2019-04-01 00:00:00', '2019-05-01 00:00:00', '2019-06-01 00:00:00'

Efectivamente, tenemos 65 valores únicos en mes_informacion, tal como en la base bd_comp2. Por otro lado, la variable comportamiento_j tiene 15 valores únicos. Así, de las bases bd_comp2 y bd_comp3 se infiere que dichas tablas toenen registro de un número de caso si este empieza un cierto trámite específico, pues no todos los numero de casos están registrados. Una vez empezado el trámite, se rastrea su comportamiento , tiempo de aparición y el mes de información.

A continuación haremos las consultas para verificar que bd_comp2 y bd_comp 3 tengan los mismos números de casos con el mismo número de registros

In [28]:
#Para la base bd_comp2
query1 ="""
       select num_caso as _id_,count(num_caso) as cantidad_apariciones
       from bd_comp2 group by num_caso
       """
num_tramites_bd_comp2=sqldf.run(query1)
num_tramites_bd_comp2.head(15)

_id_  cantidad_apariciones
0      1                    62
1      7                    29
2      8                    63
3      9                    15
4     10                    16
5     14                    56
6     18                    58
7     21                    54
8     22                    26
9     24                     4
10    25                    13
11    27                    13
12    28                    63
13    29                    57
14    31                    32

In [29]:
#Para la base bd_comp3
query2 ="""
       select num_caso as _id_,count(num_caso) as cantidad_apariciones
       from bd_comp3 group by num_caso
       """
num_tramites_bd_comp3=sqldf.run(query2)
num_tramites_bd_comp3.head(15)

_id_  cantidad_apariciones
0      1                    62
1      7                    29
2      8                    63
3      9                    15
4     10                    16
5     14                    56
6     18                    58
7     21                    54
8     22                    26
9     24                     4
10    25                    13
11    27                    13
12    28                    63
13    29                    57
14    31                    32

Efectivamente, parece ser que bd_comp2 y bd_comp3 tienen los mismo número de casos con el mismo número de registros cada uno. Lo que implica que dichas tablas se pueden unir. Adempas tmabipen observamos que tiene el mismo mes de información para cada caso. 

Por último, vamos a acomodar bd_comp3 por mes_informacion, tal como lo hicimos con bd_comp2.

In [30]:
#Acomodar bd_comp3 por mes_informacion
bd_comp3 = bd_comp3.sort_values(by ="mes_informacion")
bd_comp3.head()

num_caso      mes_informacion  comportamiento_j
32240      1220  2016-01-01 00:00:00               0.0
44912      1704  2016-01-01 00:00:00               0.0
74499      2824  2016-01-01 00:00:00               0.0
20479       775  2016-01-01 00:00:00               0.0
27608      1049  2016-01-01 00:00:00               0.0

### base_inversion

In [31]:
bd_inver.head()

_id2_  inversion
0  1758511   0.041039
1  1572947  77.422135
2  1051836  90.539953
3  1223255  90.205018
4  1894602   1.071354

In [32]:
bd_inver.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17536 entries, 0 to 17535
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   _id2_      17536 non-null  int64  
 1   inversion  17536 non-null  float64
dtypes: float64(1), int64(1)
memory usage: 274.1 KB


In [33]:
bd_inver.nunique()

_id2_        17536
inversion    17385
dtype: int64

Así, de las observaciones anteriores, se puede hacer un merge con bd_ini_int y bd_ini_ext por medio de la variable _id_.

También, nos interesa unir bd_comp 2 con bd_comp3 por medio de la variable num_caso y mes_informacion.

Por último, se uniran las bases bd_inv con bd_ini_pt1 por medio de las variable _id2_

### Análisis y agrupamiento de las bases

In [34]:
bd_ini_merged = bd_ini_int.merge(bd_ini_ext, how = "left", on = "_id_")
bd_ini_merged.head()

_id_ fecha_inicio  comp_interno1 comp_interno2  comp_interno3  \
0  2333   2016-01-04    6079.027356          b'1'           30.0   
1   488   2016-01-04     911.854103          b'0'           19.0   
2  1964   2016-01-04       0.000000          b'1'           19.0   
3  1220   2016-01-04    2482.269504          b'0'            9.0   
4  1226   2016-01-04       0.000000          b'1'            3.0   

   comp_interno4  categoria1  categoria2  categoria3  comp_externo1  \
0           50.0         2.0         9.0       175.0     324.637681   
1           48.0         3.0         3.0       164.0      15.942029   
2           53.0         4.0         3.0       164.0     144.927536   
3           70.0         3.0         9.0       154.0     727.536232   
4           75.0         3.0         9.0       148.0      30.434783   

   comp_externo2  comp_externo3  comp_externo4  
0      32.174662           91.0         -230.0  
1            NaN          132.0         -230.0  
2      14.363689           77.0         -230.0  
3      72.105717           75.0         -230.0  
4       3.016375          121.0         -230.0

In [35]:
bd_ini_merged.nunique()

_id_             50000
fecha_inicio      1557
comp_interno1      680
comp_interno2        2
comp_interno3       32
comp_interno4       55
categoria1           6
categoria2          13
categoria3          32
comp_externo1     7533
comp_externo2     6982
comp_externo3      237
comp_externo4        6
dtype: int64

In [36]:
bd_ini_merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50000 entries, 0 to 49999
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   _id_           50000 non-null  int64  
 1   fecha_inicio   50000 non-null  object 
 2   comp_interno1  50000 non-null  float64
 3   comp_interno2  50000 non-null  object 
 4   comp_interno3  50000 non-null  float64
 5   comp_interno4  50000 non-null  float64
 6   categoria1     50000 non-null  float64
 7   categoria2     50000 non-null  float64
 8   categoria3     50000 non-null  float64
 9   comp_externo1  47831 non-null  float64
 10  comp_externo2  44410 non-null  float64
 11  comp_externo3  49994 non-null  float64
 12  comp_externo4  50000 non-null  float64
dtypes: float64(10), int64(1), object(2)
memory usage: 5.3+ MB


In [37]:
bd_comp_merged = bd_comp2.groupby(["num_caso"], as_index=False)["mes_informacion"].max()
bd_comp_merged.head()

num_caso      mes_informacion
0         1  2021-05-01 00:00:00
1         7  2018-05-01 00:00:00
2         8  2021-04-01 00:00:00
3         9  2017-04-01 00:00:00
4        10  2017-05-01 00:00:00

Hacemos el groupby con el máximo sobre mes_informacion, pues nos interesa la informacion más reciente de cada cliente.

In [38]:
bd_comp_merged = bd_comp_merged.merge(bd_comp2, how = "left", on = ["num_caso", "mes_informacion"])
bd_comp_merged.head()

num_caso      mes_informacion  Tiempo_aparicion
0         1  2021-05-01 00:00:00              61.0
1         7  2018-05-01 00:00:00              28.0
2         8  2021-04-01 00:00:00              51.0
3         9  2017-04-01 00:00:00              14.0
4        10  2017-05-01 00:00:00              15.0

In [39]:
bd_comp_merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 49352 entries, 0 to 49351
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   num_caso          49352 non-null  int64  
 1   mes_informacion   49352 non-null  object 
 2   Tiempo_aparicion  49352 non-null  float64
dtypes: float64(1), int64(1), object(1)
memory usage: 1.5+ MB


In [40]:
bd_comp_merged = bd_comp_merged.merge(bd_comp3, how = "left", on = ["num_caso", "mes_informacion"])

In [41]:
bd_comp_merged.head()

num_caso      mes_informacion  Tiempo_aparicion  comportamiento_j
0         1  2021-05-01 00:00:00              61.0               0.0
1         7  2018-05-01 00:00:00              28.0               0.0
2         8  2021-04-01 00:00:00              51.0               0.0
3         9  2017-04-01 00:00:00              14.0               0.0
4        10  2017-05-01 00:00:00              15.0               0.0

In [42]:
bd_comp_merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 49352 entries, 0 to 49351
Data columns (total 4 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   num_caso          49352 non-null  int64  
 1   mes_informacion   49352 non-null  object 
 2   Tiempo_aparicion  49352 non-null  float64
 3   comportamiento_j  49352 non-null  float64
dtypes: float64(2), int64(1), object(1)
memory usage: 1.9+ MB


In [43]:
bd_inv_merged = bd_ini_pt1.merge(bd_inver, how = "left", on = "_id2_")
bd_inv_merged.head()

_id_    _id2_  comp_mixto1  inversion
0  78213  1798899      0.33522        NaN
1  76457  1758511      0.33334   0.041039
2  40666   935318      0.29186        NaN
3   8329   191567      0.29112        NaN
4  68389  1572947      0.27318  77.422135

In [44]:
bd_inv_merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50000 entries, 0 to 49999
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   _id_         50000 non-null  int64  
 1   _id2_        50000 non-null  int64  
 2   comp_mixto1  50000 non-null  float64
 3   inversion    17536 non-null  float64
dtypes: float64(2), int64(2)
memory usage: 1.9 MB


In [45]:
bd_inv_merged.nunique()

_id_           50000
_id2_          50000
comp_mixto1     7098
inversion      17385
dtype: int64

### Veamos en dónde y cuales id's coinciden según la base.

In [46]:
x = set(bd_ini_pt1["_id_"])
print(len(x))
y = set(bd_ini_pt1["_id2_"])
print(len(y))
z = set(bd_comp2["num_caso"])
print(len(z))
w = set(bd_comp3["num_caso"])
print(len(w))
u = set(bd_ini_int["_id_"])
print(len(u))
v = set(bd_ini_ext["_id_"])
print(len(v))

50000
50000
49352
49352
50000
50000


In [47]:
x_y = y.intersection(x)
print(len(x_y))
x_z = x.intersection(z)
print(len(x_z))
x_w = x.intersection(w)
print(len(x_w))
y_z = y.intersection(z)
print(len(y_z))
y_w = y.intersection(w)
print(len(y_w))
z_w = z.intersection(w)
print(len(z_w))
u_v = u.intersection(v)
print(len(u_v))

1069
49352
49352
1053
1053
49352
50000


De lo siguiente podemos observar que, en gran medida, la columna _id_ de bd_ini_pt1 coincide con num_caso de ambas bases de comportamiento, por lo que cruce de esas bases con bd_ini_pt1 debería ser con la columna _id_. De la misma manera, podemos observar que la columna _id_ coincide en las bases externa e interna. Veamos si estas mismas coincide con el _id_ de bd_ini_pt1.

In [48]:
u_v_x = u_v.intersection(x)
len(u_v_x)

50000

Efectivamente, estos id's coinciden, por lo que el agrupamiento de estas bases debería ser sobre la columna id.

In [49]:
bd_id = bd_ini_pt1.merge(bd_ini_int, how = "left", on = "_id_")
bd_id = bd_id.sort_values(by = "fecha_inicio")
bd_id.head()

_id_  _id2_  comp_mixto1 fecha_inicio  comp_interno1 comp_interno2  \
17236  2333  53659      0.32856   2016-01-04    6079.027356          b'1'   
47365   488  11224      0.38498   2016-01-04     911.854103          b'0'   
21961  1964  45172      0.34780   2016-01-04       0.000000          b'1'   
49384  1220  28060      0.28422   2016-01-04    2482.269504          b'0'   
16983  1226  28198      0.31500   2016-01-04       0.000000          b'1'   

       comp_interno3  comp_interno4  categoria1  categoria2  categoria3  
17236           30.0           50.0         2.0         9.0       175.0  
47365           19.0           48.0         3.0         3.0       164.0  
21961           19.0           53.0         4.0         3.0       164.0  
49384            9.0           70.0         3.0         9.0       154.0  
16983            3.0           75.0         3.0         9.0       148.0

In [50]:
bd_id.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50000 entries, 17236 to 32134
Data columns (total 11 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   _id_           50000 non-null  int64  
 1   _id2_          50000 non-null  int64  
 2   comp_mixto1    50000 non-null  float64
 3   fecha_inicio   50000 non-null  object 
 4   comp_interno1  50000 non-null  float64
 5   comp_interno2  50000 non-null  object 
 6   comp_interno3  50000 non-null  float64
 7   comp_interno4  50000 non-null  float64
 8   categoria1     50000 non-null  float64
 9   categoria2     50000 non-null  float64
 10  categoria3     50000 non-null  float64
dtypes: float64(7), int64(2), object(2)
memory usage: 4.6+ MB


In [51]:
bd_id.nunique()

_id_             50000
_id2_            50000
comp_mixto1       7098
fecha_inicio      1557
comp_interno1      680
comp_interno2        2
comp_interno3       32
comp_interno4       55
categoria1           6
categoria2          13
categoria3          32
dtype: int64

In [52]:
bd_final = bd_id.merge(bd_inv_merged, how = "inner", on = ["_id2_", "_id_", "comp_mixto1"])
bd_final.head()

_id_  _id2_  comp_mixto1 fecha_inicio  comp_interno1 comp_interno2  \
0  2333  53659      0.32856   2016-01-04    6079.027356          b'1'   
1   488  11224      0.38498   2016-01-04     911.854103          b'0'   
2  1964  45172      0.34780   2016-01-04       0.000000          b'1'   
3  1220  28060      0.28422   2016-01-04    2482.269504          b'0'   
4  1226  28198      0.31500   2016-01-04       0.000000          b'1'   

   comp_interno3  comp_interno4  categoria1  categoria2  categoria3  inversion  
0           30.0           50.0         2.0         9.0       175.0        NaN  
1           19.0           48.0         3.0         3.0       164.0        NaN  
2           19.0           53.0         4.0         3.0       164.0        NaN  
3            9.0           70.0         3.0         9.0       154.0        NaN  
4            3.0           75.0         3.0         9.0       148.0   0.311552

In [53]:
bd_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50000 entries, 0 to 49999
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   _id_           50000 non-null  int64  
 1   _id2_          50000 non-null  int64  
 2   comp_mixto1    50000 non-null  float64
 3   fecha_inicio   50000 non-null  object 
 4   comp_interno1  50000 non-null  float64
 5   comp_interno2  50000 non-null  object 
 6   comp_interno3  50000 non-null  float64
 7   comp_interno4  50000 non-null  float64
 8   categoria1     50000 non-null  float64
 9   categoria2     50000 non-null  float64
 10  categoria3     50000 non-null  float64
 11  inversion      17536 non-null  float64
dtypes: float64(8), int64(2), object(2)
memory usage: 5.0+ MB


In [54]:
bd_final.sort_values(by = "_id_")

_id_    _id2_  comp_mixto1 fecha_inicio  comp_interno1 comp_interno2  \
1864       1       23      0.38000   2016-03-31    3039.513678          b'1'   
238        7      161      0.33360   2016-01-19    1547.112462          b'1'   
718        8      184      0.27898   2016-02-10    1215.805471          b'1'   
909        9      207      0.38000   2016-02-18       0.000000          b'1'   
914       10      230      0.37050   2016-02-18       0.000000          b'1'   
...      ...      ...          ...          ...            ...           ...   
49735  98987  2276701      0.33712   2021-04-13       0.000000          b'1'   
49764  98988  2276724      0.26010   2021-04-14    2026.342452          b'1'   
49996  98990  2276770      0.30434   2021-04-29       0.000000          b'1'   
49992  98992  2276816      0.30506   2021-04-29       0.000000          b'1'   
49904  98997  2276931      0.35322   2021-04-21       0.000000          b'0'   

       comp_interno3  comp_interno4  categoria1  categoria2  categoria3  \
1864            26.0           67.0         3.0         3.0       171.0   
238             28.0           75.0         1.0        11.0       173.0   
718             14.0           55.0         2.0         6.0       159.0   
909             11.0           55.0         2.0         9.0       156.0   
914             11.0           56.0         2.0         6.0       156.0   
...              ...            ...         ...         ...         ...   
49735           23.0           51.0         3.0         9.0       168.0   
49764           19.0           63.0         3.0         9.0       164.0   
49996           17.0           60.0         2.0         9.0       162.0   
49992           15.0           47.0         1.0         9.0       160.0   
49904           15.0           43.0         1.0         9.0       160.0   

         inversion  
1864     52.500490  
238       0.047480  
718     227.537211  
909       0.921995  
914      39.259568  
...            ...  
49735          NaN  
49764  2596.865519  
49996   107.495449  
49992     9.822964  
49904          NaN  

[50000 rows x 12 columns]

De esta manera sólo nos falta agrupar bd_comp_merged y bd_final

In [55]:
bd_comp_merged.head()

num_caso      mes_informacion  Tiempo_aparicion  comportamiento_j
0         1  2021-05-01 00:00:00              61.0               0.0
1         7  2018-05-01 00:00:00              28.0               0.0
2         8  2021-04-01 00:00:00              51.0               0.0
3         9  2017-04-01 00:00:00              14.0               0.0
4        10  2017-05-01 00:00:00              15.0               0.0

In [56]:
bd_comp_merged.rename(columns={"num_caso": "_id_"},inplace=True)

In [57]:
bd_comp_merged.head()

_id_      mes_informacion  Tiempo_aparicion  comportamiento_j
0     1  2021-05-01 00:00:00              61.0               0.0
1     7  2018-05-01 00:00:00              28.0               0.0
2     8  2021-04-01 00:00:00              51.0               0.0
3     9  2017-04-01 00:00:00              14.0               0.0
4    10  2017-05-01 00:00:00              15.0               0.0

In [58]:
bd_final = bd_final.merge(bd_comp_merged, how = "left", on = "_id_")

In [59]:
bd_final.head()

_id_  _id2_  comp_mixto1 fecha_inicio  comp_interno1 comp_interno2  \
0  2333  53659      0.32856   2016-01-04    6079.027356          b'1'   
1   488  11224      0.38498   2016-01-04     911.854103          b'0'   
2  1964  45172      0.34780   2016-01-04       0.000000          b'1'   
3  1220  28060      0.28422   2016-01-04    2482.269504          b'0'   
4  1226  28198      0.31500   2016-01-04       0.000000          b'1'   

   comp_interno3  comp_interno4  categoria1  categoria2  categoria3  \
0           30.0           50.0         2.0         9.0       175.0   
1           19.0           48.0         3.0         3.0       164.0   
2           19.0           53.0         4.0         3.0       164.0   
3            9.0           70.0         3.0         9.0       154.0   
4            3.0           75.0         3.0         9.0       148.0   

   inversion      mes_informacion  Tiempo_aparicion  comportamiento_j  
0        NaN  2021-05-01 00:00:00              52.0               0.0  
1        NaN  2021-05-01 00:00:00              64.0               0.0  
2        NaN  2021-05-01 00:00:00              64.0               0.0  
3        NaN  2021-05-01 00:00:00              58.0               0.0  
4   0.311552  2018-10-01 00:00:00              33.0               0.0

In [60]:
bd_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50000 entries, 0 to 49999
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   _id_              50000 non-null  int64  
 1   _id2_             50000 non-null  int64  
 2   comp_mixto1       50000 non-null  float64
 3   fecha_inicio      50000 non-null  object 
 4   comp_interno1     50000 non-null  float64
 5   comp_interno2     50000 non-null  object 
 6   comp_interno3     50000 non-null  float64
 7   comp_interno4     50000 non-null  float64
 8   categoria1        50000 non-null  float64
 9   categoria2        50000 non-null  float64
 10  categoria3        50000 non-null  float64
 11  inversion         17536 non-null  float64
 12  mes_informacion   49352 non-null  object 
 13  Tiempo_aparicion  49352 non-null  float64
 14  comportamiento_j  49352 non-null  float64
dtypes: float64(10), int64(2), object(3)
memory usage: 6.1+ MB


In [61]:
bd_final.nunique()

_id_                50000
_id2_               50000
comp_mixto1          7098
fecha_inicio         1557
comp_interno1         680
comp_interno2           2
comp_interno3          32
comp_interno4          55
categoria1              6
categoria2             13
categoria3             32
inversion           17385
mes_informacion        62
Tiempo_aparicion       65
comportamiento_j       15
dtype: int64

_De esta manera, nuestra base final es bd_final_

In [62]:
bd_final.head(20)

_id_  _id2_  comp_mixto1 fecha_inicio  comp_interno1 comp_interno2  \
0   2333  53659      0.32856   2016-01-04    6079.027356          b'1'   
1    488  11224      0.38498   2016-01-04     911.854103          b'0'   
2   1964  45172      0.34780   2016-01-04       0.000000          b'1'   
3   1220  28060      0.28422   2016-01-04    2482.269504          b'0'   
4   1226  28198      0.31500   2016-01-04       0.000000          b'1'   
5   1278  29394      0.36858   2016-01-04       0.000000          b'1'   
6    899  20677      0.26702   2016-01-04       0.000000          b'1'   
7   2235  51405      0.38000   2016-01-04       0.000000          b'0'   
8    786  18078      0.35078   2016-01-04       0.000000          b'0'   
9   2113  48599      0.37506   2016-01-04       0.000000          b'1'   
10  2410  55430      0.32920   2016-01-04       0.000000          b'0'   
11   277   6371      0.35336   2016-01-04     658.561297          b'1'   
12  2383  54809      0.37110   2016-01-04    2026.342452          b'1'   
13  1468  33764      0.34000   2016-01-04       0.000000          b'1'   
14   924  21252      0.38000   2016-01-04    1519.756839          b'1'   
15    64   1472      0.38000   2016-01-05     709.219858          b'1'   
16  1152  26496      0.33548   2016-01-05       0.000000          b'1'   
17  2421  55683      0.37000   2016-01-05    1114.488349          b'1'   
18  1816  41768      0.37000   2016-01-05    3039.513678          b'1'   
19  1935  44505      0.29150   2016-01-05    1013.171226          b'0'   

    comp_interno3  comp_interno4  categoria1  categoria2  categoria3  \
0            30.0           50.0         2.0         9.0       175.0   
1            19.0           48.0         3.0         3.0       164.0   
2            19.0           53.0         4.0         3.0       164.0   
3             9.0           70.0         3.0         9.0       154.0   
4             3.0           75.0         3.0         9.0       148.0   
5            19.0           45.0         3.0         9.0       164.0   
6             1.0           60.0         1.0         9.0       146.0   
7            15.0           73.0         1.0        11.0       160.0   
8            23.0           68.0         3.0         9.0       168.0   
9            19.0           44.0         3.0         9.0       164.0   
10           15.0           72.0         1.0         5.0       160.0   
11           19.0           57.0         2.0         9.0       164.0   
12           15.0           59.0         1.0         9.0       160.0   
13           19.0           72.0         2.0        11.0       164.0   
14            1.0           53.0         2.0         9.0       146.0   
15           19.0           60.0         1.0         9.0       164.0   
16           14.0           46.0         3.0         9.0       159.0   
17           17.0           52.0         1.0         9.0       162.0   
18           24.0           65.0         1.0         6.0       169.0   
19           21.0           69.0         1.0         9.0       166.0   

     inversion      mes_informacion  Tiempo_aparicion  comportamiento_j  
0          NaN  2021-05-01 00:00:00              52.0               0.0  
1          NaN  2021-05-01 00:00:00              64.0               0.0  
2          NaN  2021-05-01 00:00:00              64.0               0.0  
3          NaN  2021-05-01 00:00:00              58.0               0.0  
4     0.311552  2018-10-01 00:00:00              33.0               0.0  
5          NaN  2018-08-01 00:00:00              31.0               0.0  
6          NaN  2021-05-01 00:00:00              63.0               0.0  
7          NaN  2020-12-01 00:00:00              58.0               1.0  
8          NaN  2018-02-01 00:00:00              24.0               0.0  
9          NaN  2021-05-01 00:00:00              64.0               1.0  
10         NaN  2021-05-01 00:00:00              63.0               0.0  
11         NaN  2019-07-01 00:00:00             